src/compiler/checker.ts

```diff
@@ -24307,6 +24321,68 @@ namespace ts {
                 errorOutputContainer);
+
+            function checkTagNameDoesNotExpectTooManyArguments(): boolean {
+                const tagType = isJsxOpeningElement(node) || isJsxSelfClosingElement(node) && !isJsxIntrinsicIdentifier(node.tagName) ? checkExpression(node.tagName) : undefined;
+                if (!tagType) {
+                    return true;
+                }
+                const tagCallSignatures = getSignaturesOfType(tagType, SignatureKind.Call);
+                if (!length(tagCallSignatures)) {
+                    return true;
+                }
+                const factory = getJsxFactoryEntity(node);
+                if (!factory) {
+                    return true;
+                }
+                const factorySymbol = resolveEntityName(factory, SymbolFlags.Value, /*ignoreErrors*/ true, /*dontResolveAlias*/ false, node);
+                if (!factorySymbol) {
+                    return true;
+                }
+
+                const factoryType = getTypeOfSymbol(factorySymbol);
+                const callSignatures = getSignaturesOfType(factoryType, SignatureKind.Call);
+                if (!length(callSignatures)) {
+                    return true;
+                }
+
+                let hasFirstparamSignatures = false;
+                // Check that _some_ first parameter expects and SFC-like thing, and that some overload of the SFC expects an acceptable number of arguments
+                for (const sig of callSignatures) {
+                    const firstparam = getTypeAtPosition(sig, 0);
+                    const signaturesOfParam = getSignaturesOfType(firstparam, SignatureKind.Call);
+                    if (!length(signaturesOfParam)) continue;
+                    for (const paramSig of signaturesOfParam) {
+                        hasFirstparamSignatures = true;
+                        if (hasEffectiveRestParameter(paramSig)) {
+                            return true; // some signature has a rest param, so function components can have an aritrary number of arguments
+                        }
+                        const paramCount = getParameterCount(paramSig);
+                        for (const tagSig of tagCallSignatures) {
+                            const tagParamCount = getParameterCount(tagSig);
+                            if (tagParamCount <= paramCount) {
+                                return true; // some signature accepts the number of arguments the function component provides
+                            }
+                        }
+                    }
+                }
+                if (!hasFirstparamSignatures) {
+                    // Not a single signature had a first parameter which expected a signature - for back compat, and
+                    // to guard against generic factories which won't have signatures directly, return true
+                    return true;
+                }
+
+                if (reportErrors) {

```

Comment:
shouldn't we just skip this whole function if reportErrors is false?
Edit: Oh, I guess that `false` from this function skips the checkTypeRelatedTo call.

Reply:
Yeah, `reportErrors` doesn't outright disable relationship checking, it just makes it so we don't -_report_ errors :P